lib/pacemaker/libpacemaker_private.h

```diff
@@ -25,8 +25,21 @@ struct resource_alloc_functions_s {
     gboolean(*create_probe) (pe_resource_t *, pe_node_t *, pe_action_t *, gboolean, pe_working_set_t *);
     void (*internal_constraints) (pe_resource_t *, pe_working_set_t *);
 
+    /*!
+     * \internal
+     * \brief Apply a colocation's score to node weights or resource priority
+     *
+     * Given a colocation constraint, apply its score to the dependent's
+     * allowed node weights (if we are still placing resources) or priority (if
+     * we are choosing promotable clone instance roles).
+     *
+     * \param[in] dependent      Dependent resource in colocation
+     * \param[in] primary        Primary resource in colocation
```

Comment:
Would one or both of these be [in,out] since we're updating the resource's nodes table?

Reply:
It's a good question, we've never decided how `out` should apply to struct pointers, and I haven't seen anything about it in doxygen documentation either. Should a struct pointer s be `out` only if any member of s is changed (shallow, e.g. s->s2), or if anything that can be reached via s is changed (deep, e.g. s->s2->member)? Or in this case, replacing dependent->nodes altogether vs. changing the contents of the existing hash table pointed to by dependent->nodes.

I've addressed the question so far by ignoring it :)

Reply:
Related, I have often make pcmk__output_t an [in,out] because things like the XML formatter are building up out->priv->root by adding stuff to that tree.  So, there's precedence if you want to use it.